docs/proposals/15-manage-bastions-and-ssh-key-pair-rotation.md

```diff
@@ -0,0 +1,201 @@
+The static shoot-specific SSH key pair should be rotated regularily, for example once in the maintenance time window.
+
+In a [previous proposal](https://github.com/gardener/gardenctl/issues/508) it was suggested that an extension, running on the seed cluster, watches `Bastion` custom resources on the garden cluster, acting upon it and updating it's status accordingly.
+However changes to the `Bastion` resource should only be allowed for controllers on seeds that are responsible for it. This cannot be restricted when using custom resources.
+The proposal, as outlined below, suggests to implement the necessary changes in the gardener core components and to adapt the [SeedAuthorizer](https://github.com/gardener/gardener/issues/1723) to consider `Bastion` resources that the Gardener API Server serves.
+
+
+### Goals
+- Operators have own SSH keys to access the bastion host
+- The SSH private key for the bastion host is encrypted with the user's public PGP key, so that only the user who requested the SSH session can access the bastion
+- A controller on the seed ensures that the temporary infrastructure resources for the bastion are cleaned up after afterwards
+
+### Non-Goals
+- Restricting access to specific node groups and to have temporary OS users and SSH keys for a user/operator on the shoot worker nodes (for auditing purposes). We had the impression that the effort doesn't justify the security improvement (at least for now as the first step; we might reconsider this at a later point in time).
+- Shared bastion host, in case multiple operators want to access the same shoot
+- Use machine controller manager to create the bastion
+
+## Proposal
+
+### Involved Components
+The following is a list of involved components, that either need to be newly introduced or extended if already existing
+- `gardenctlv2` (or any other client)
+    - Creates `Bastion` resource
+    - SSH to shoot node through a bastion
+    - Heartbeats / keeps alive the `Bastion` resource during SSH connection
+- Gardener extension provider <infra>
+    - Provider specific bastion controller
+    - Should be added to gardener-extension-provider-<infra> repos, e.g. https://github.com/gardener/gardener-extension-provider-aws/tree/master/pkg/controller
+    - Has the permission to update the `Bastion/status` subresource on the seed cluster
+    - Runs on seed (of course)
+- Gardener API Server (`GAPI`)
+    - New `operations.gardener.cloud` API Group
+    - New resource type `Bastion`
+    - New Admission Webhooks for `Bastion` resource
+    - `SeedAuthorizer`: The `SeedAuthorizer` and dependency graph needs to be extended to consider the `Bastion` resource https://github.com/gardener/gardener/tree/master/pkg/admissioncontroller/webhooks/auth/seed/graph
```

Comment:
Personally, I prefer `Bastion`, because this is what will be materialized out of this resource. If you/gardenctl/... use(s) it for SSH or for whatever else is up to you and should not be encoded in the resource name.

Reply:
> Wouldn't it have been better to start with this first (helping the audience understand your proposal)?

In the motivation I wrote:

> [...] However changes to the `Bastion` resource should only be allowed for controllers on seeds that are responsible for it. This cannot be restricted when using custom resources.
The proposal, as outlined below, suggests to implement the necessary changes in the gardener core components and to adapt the [SeedAuthorizer](https://github.com/gardener/gardener/issues/1723) to consider `Bastion` resources that the Gardener API Server serves.

Isn't this what you were looking for?

> And why is it a bastion resource in the GAPI, not an ssh (or whatever) resource?

My initial proposal was to call it `SSH` and afterwards renamed it to `Bastion` after the suggestion from @rfranzke. I do not have a strong opinion on this..

Reply:
@petersutter What I meant was the list itself, i.e. "Involved Components". I would find it more helpful to start with that happens first in the Garden cluster.

Reply:
@petersutter @rfranzke For what it's worth, I very strongly lean towards `ssh`, because that's what the operator requests here and everything else is only enabling that goal (e.g. the bastion). As an operator, I do not want a bastion (and may not need it in all cases), but I want `ssh` access to the nodes. If we call it bastion, we corner us into this particular implementation detail.

@rfranzke What's your reasoning for bastion?

Reply:
See [above](https://github.com/gardener/gardener/pull/3802#discussion_r609481086), my understanding is that the controllers inside Gardener/extensions have nothing to do with the SSH session as such. They only prepare the infrastructure and materialize the VM, but the SSH session itself is out of scope for them and only established later by the client.
In this case, it's rather only an implementation detail of `gardenctl` (one of potentially many clients of this new API) because `gardenctl ssh` creates a new SSH connection with the help of such bastion. There might be other applications of this API in the future, so if there is no strong need to encode "ssh" details then I would still prefer to name it in a general way and keep `Bastion`.
We also call it `Terminal` instead of `KubectlAccess` (although, that's what I eventually want as operator) or `Machine` instead of `Kubelet` (although, that's what I eventually want as gardenlet).

Reply:
@rfranzke I need to think it through more. But just as a basic rebuttal/counter-argument:
- On the one hand: Most prominently, we do create `shoot` resources and expect them to be created (in this analogy: `ssh`), not to create the building blocks like a control plane and the nodes ourselves (in this analogy: `bastion`), which is/are the building blocks.
- On the other hand: The `machine` vs. `kubelet` argument is actually one speaking in favour of the `ssh` argument, isn't it? I want a `machine` that is more than the `node`, the `kubelet`, etc. so I say `machine` (in this analogy: `ssh`) and not `node`, `kubelet`, etc. (in this analogy: `bastion`), which are the building blocks.
- And as for terminals: We do want web terminals, not `kubectl` access. We have that already. So, the main goal were indeed web terminals. They were no means to an end - they were what we wanted to achieve/get/have.

Anyway, give me some time, but my gut feeling has not changed. We talked out-of-band about the need or no need for a more complex solution and agreed to a simpler solution (no bastion reuse, no node specific access, etc.), but you and @petersutter said that we may add that later. This also speaks to the notion of dealing with `ssh` rather than focusing on `bastion` as a resource, but it can be changed later of course, too.

I guess, even if we stick to `bastion`, I would "frame" and "motivate" the GEP nonetheless with the goal of `ssh`, because that's what we fundamentally want.

Reply:
BTW, are we always having a bastion host? What are we doing on GCP? Wasn't it the case previously to "only" modify the firewall rules? In other words, will there always be a bastion on all infrastructures?

Reply:
Yeah, well, that's indeed a good point which speaks against `Bastion` and some different name, true...

Reply:
In a chat with @petersutter and @xrstf we decided that we always (for all infrastructures) want to provision a dedicated (bastion) VM, even if there are easier/more direct ways to get access (e.g., on GCP/Azure it can be enough to open a dedicated firewall rule and attach a public IP to an existing node).

If we would try to re-use existing nodes (e.g., on GCP/Azure) then we might get new complexities with transferring the personalized, short-living RSA keys to the shoot worker nodes. It's even unclear which node is chosen and whether this works for all use-cases in which an operator wants/needs to get access.

Overall, it seems easier and more harmonized across all IaaS to create dedicated bastion VMs, hence, the resource name `Bastion` seems appropriate and we would like to keep it.

Reply:
...or we don't and keep the Azure and GCP cases simple/without bastion.

We have cut down the proposal already so much, and given the additional effort on Azure and GCP, why not drop also the generated `ssh` keys for the bastion.

We have:
- Node `ssh` keys are generated and rotated by Gardener anyhow
- We have to build client IP allow-listing anyhow

We can decide that this is already sufficient and if we don't have bastions on Azure and GCP and the operator uses his own keypair to access the bastion on AWS, then let's do that and simplify it even more.

Just a thought. As said, we already have node `ssh` keys and client IP allow-listing. If that's sufficient, let's draw the line there if the original scope is too much at first.